### ISSUE TITLE
[Test failed] ilasm_System._Runtime_CompilerServices_MethodImplOptionsTests_MethodImplOptionsTests_._Runtime_CompilerServices_MethodImplOptionsTests_MethodImplOptionsTests_sh

### ISSUE THREAD

**Issue by VincentBu** [dotnet/coreclr#25067]
**Job:**
https://mc.dot.net/#/user/coreclr-outerloop-jitstress/ci~2Fdotnet~2Fcoreclr~2Frefs~2Fheads~2Fmaster/test~2Ffunctional~2Fcli~2F/20190609.1/workItem/PayloadGroup0/analysis/xunit/ilasm_System._Runtime_CompilerServices_MethodImplOptionsTests_MethodImplOptionsTests_~2F_Runtime_CompilerServices_MethodImplOptionsTests_MethodImplOptionsTests_sh

**Failed tests:**
ilasm_System._Runtime_CompilerServices_MethodImplOptionsTests_MethodImplOptionsTests_._Runtime_CompilerServices_MethodImplOptionsTests_MethodImplOptionsTests_sh

**Log:**
```
Stack overflow.
/root/helix/work/workitem/ilasm/System/Runtime/CompilerServices/MethodImplOptionsTests/MethodImplOptionsTests.sh: line 252: 10806 Aborted                 (core dumped) $LAUNCHER $ExePath \"${CLRTestExecutionArguments[@]}\"

Return code:      1
Raw output file:      /root/helix/work/workitem/ilasm/System/Reports/ilasm.System/Runtime/CompilerServices/MethodImplOptionsTests/MethodImplOptionsTests.output.txt
Raw output:
BEGIN EXECUTION
/root/helix/work/correlation/corerun MethodImplOptionsTests.exe ''
AggressiveOptimizationTest

Assembling 'MiAggressiveOptimization.il'  to DLL --> 'MiAggressiveOptimization.dll'
Source file is ANSI

Assembled method Program::Main
Creating PE file

Emitting classes:
Class 1:\tProgram

Emitting fields and methods:
Global \t
Class 1\tMethods: 1;\t

Emitting events and properties:
Global \t
Class 1\t
Writing PE file
Operation completed successfully
Expected: 100
Actual: 134
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=/root/helix/work/correlation
> /root/helix/work/workitem/ilasm/System/Runtime/CompilerServices/MethodImplOptionsTests/MethodImplOptionsTests.sh
Expected: True
Actual:   False
```

**Comment by BruceForstall** [comment]
Looks like this test failed in a number of different configurations:
```
Alpine.38.Arm64.Open-arm64-Checked-zapdisable
Debian.9.Amd64.Open-x64-Checked-gcstress0xc_zapdisable_heapverify1
Ubuntu.1604.Amd64.Open-x64-Checked-gcstress0xc_zapdisable_heapverify1
Debian.9.Arm64.Open-arm64-Checked-gcstress0xc_zapdisable_heapverify1
Ubuntu.1804.Arm32.Open-arm-Checked-gcstress0xc_zapdisable_heapverify1
Ubuntu.1804.Arm64.Open-arm64-Checked-gcstress0xc_zapdisable_heapverify1
Ubuntu.1804.Arm32.Open-arm-Checked-gcstress0x3
```

**Comment by sandreenko** [comment]
JIT/superpmi/superpmicollect/superpmicollect.sh fails with the same symptoms.

I will take a look at what do we have on the stack.

**Comment by sandreenko** [comment]
Ok, here what I got:
1. both of them fail with StackOverflow, it happens only on Alpine, doesn't repro on normal Ubuntu (even when stack size is set to 1.5Mb as it is on Alpine or even smaller);

2. when the StackOverflow happens we have 60 jit frames on the top of the stack, most of them are fgMorph*, each `Compiler::fgMorphSmpOp` has 5kb size frame.

3. this size is wasted on temp variables like these:
```
   0x00007ffff4cfe88e <+40990>: mov    QWORD PTR [rbp-0x1568],rax <- the only def
   0x00007ffff4cfe895 <+40997>: jne    0x7ffff4cfe8b3 <Compiler::fgMorphSmpOp(GenTree*, Compiler::MorphAddrContext*)+41027>
   0x00007ffff4cfe89b <+41003>: mov    rax,QWORD PTR [rbp-0x1568] <- the only use
```
and there are 1500 of such vars:
```
Dump of assembler code for function Compiler::fgMorphSmpOp(GenTree*, Compiler::MorphAddrContext*):
   0x00007ffff4cf4870 <+0>:     push   rbp
   0x00007ffff4cf4871 <+1>:     mov    rbp,rsp
   0x00007ffff4cf4874 <+4>:     sub    rsp,0x1570
```
on Windows there is no such problem, the frame there as small as it expected to be.

4. I was not able to get the managed stack (tried https://github.com/dotnet/diagnostics/blob/master/documentation/dotnet-dump-instructions.md , thanks @janvorli  for the advice), but it was not able to create a core dump.

I do not think that 3. is the main problem here (because normal Linux doesn't repro it), but it is something that would be nice to have fixed (update clang? change its options?).

I will continue investigating.

cc @dotnet/jit-contrib

**Comment by janvorli** [comment]
> 1. it was not able to create a core dump.

I am sorry for not being clear. I was not referring to creating a core dump using that tool, but rather debugging a dump created by the OS using `dotnet-dump analyze`. @hoyosjs is there a known problem with Alpine Linux and the dotnet-dump tool?

**Comment by hoyosjs** [comment]
@janvorli Not sure, according to to Mike's testing the analyze command should be working as expected. After talking to @sandreenko it looks like it can look at the managed stack, but it doesn't solve his issue.

**Comment by sandreenko** [comment]
@hoyosjs helped me to make `dotnet-dump analyze` working.

It gave me the top managed frames:
```
        Child SP               IP Call Site
00007FED483E7538 00007fed4de68a50 [PrestubMethodFrame: 00007fed483e7538] System.Collections.Generic.Dictionary`2[[System.Int32, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].FindEntry(Int32)
00007FED483E7690 00007FECD3C21DD7 System.Collections.Generic.Dictionary`2[[System.Int32, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryGetValue(Int32, System.__Canon ByRef)
00007FED483E76B0 00007FECD3C219C7 System.Diagnostics.ProcessWaitState.CheckChildren(Boolean) [/_/src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs @ 606]
00007FED483E7780 00007FECD3C20E2E System.Diagnostics.Process.OnSigChild(Boolean) [/_/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs @ 1027]
00007FED483E77C0 00007FECD3C20DB0 ILStubClass.IL_STUB_ReversePInvoke(Int32)
```

so now we know top 60 native frames (that take 80kb) and 3 managed frames before them (that take ~1kb).

Interesting that I can repro this issue on normal Ubuntu when I set stack limit to 120kb (via ulimit -S -s 120). @janvorli do you remember how we change stack size for all clr threads there? I suspect that maybe we do not set it to 1.5Mb but to 150Kb instead?

**Comment by sandreenko** [comment]
We finally found the root cause, it was the thread that did not get 1.5Mb stack because it was created by CoreFX ([code is living here](https://github.com/dotnet/corefx/blob/007571b90c3cc555cd30f3b467b87c22f9f665e0/src/Native/Unix/System.Native/pal_signal.c#L242-L286)).

@janvorli has a fix for it.

For future issues like this: the fastest way to debug this issue was to print id of each thread where we set "pthread_attr_setstacksize" to 1.5Mb, then confirm that the thread that hits overflow doesn't appear in these logs, then set a breakpoint on `__pthread_create` and wait for this thread to hit this breakpoint, get bt there and understand why it skipped `pthread_attr_setstacksize`).
